vpr/src/base/atom_netlist_utils.cpp

```diff
@@ -471,7 +471,7 @@ int mark_undriven_primitive_outputs_as_constant(AtomNetlist& netlist, int verbos
                 }
             }
 
-            if (!has_connected_inputs) {
+            if (!has_connected_inputs && !upstream_ports.empty()) {
```

Comment:
This needs a comment explaining what !upstream_ports.empty() indicates (why is this not a constant generator if none of its inputs are connected to anything?). An example would be useful.

Reply:
Wouldn't a black box still have some inputs of the atom connected (so upstream_ports wouldn't necessarily be empty)?  I am likely missing the point, but that means a comment with an example is needed I think.